docs/design/2022-03-25-non-transactional-statement.md

```diff
@@ -0,0 +1,123 @@
+
+Different from the deprecated batch-DML, a non-transactional DML splits SQLs so every split SQL is a normal statement and does not risk data-index consistency.
+
+## Detailed Design
+
+### Syntax
+
+The syntax: `split on <column_name> limit <batch_size> <DML>`
+
+In the first step only `delete` is going to be supported, but `update` and `insert into select` are also worth considering.
+
+There is no other constraints on the DML.
+
+There can be a dry run syntax `dryrun split on <column_name> limit <batch_size> <DML>` that just output how the statement will be split and an representative of their query plans.
+
+### Session
+
+Different from most of the statements, the non-transactional DML is handled in the session level. 
+
+Non-transactional statements are treated as `SimplePlans` and we won't compile it. In `runStmt`, instead of `s.Exec` for most plans, `handleNontransactionalDML` will be called. In `handleNontransactionalDML`, we will split the DML into multiple statements and execute them one by one using `Session.ExecuteStatement`.
+
+### How the split work
+
+To find the split keys, a `select` is used to read the split column specified by the user. For example, for `split on a limit 1000 delete from t where b < 1000`, the select statement would look like: `select a from t where b < 1000 order by a`.
+The result set could be large, but we don't need all of them, so result chunks are batched until the size of the batch is greater than or equal to the specified `batchSize`. Only the first and last element of each batch are kept, and thus forming a job. A job specifies a range in the specified column.
+
+Theoretically jobs can be executed in parallel, but that would require multiple sessions to execute them. There are two kinds of sessions in TiDB: user sessions that must be bound to client connections in a 1-on-1 manner, and internal sessions. Both kinds of sessions are inappropriate the split statements, so in favor of maintainence, only the current user session will be used to execute the jobs, which is serial.
+
+Each split statement is generated by embedding the split range in where clause of the given DML using a between operator.
+For example, the original statement is `split on a limit 1000 delete from t where b < 1000`, a job `{start:1, end:1000}` generates a split SQL `from t where (a between 1 and 1000) and b < 1000`.
+
+### Error handling
+
+A non-transactional statement obviously cannot rollback, so when one of the split statements fails, it just collects the error and continue on the job until all jobs are finished. At last, a set of errors (if there are) are returned to the user. Users may decide how to deal with these errors.
+
+If the statement is aborted by user, it should report its progress and return the errors it has collected.
```

Comment:
Is there a way to get the progress of a running non-transactional DML?

Reply:
The ways I can think of:
- Logs, we already have them
- Writing the progress to a system table
- A heatmap in Grafana?
- Update the progress in the processInfo.

Any suggestions?

Reply:
Logs and progress in the processinfo looks good to me.